src/Microsoft.AspNet.OData.Shared/Builder/ODataConventionModelBuilder.cs

```diff
@@ -1080,21 +1080,48 @@ private void ApplyPropertyConvention(IEdmPropertyConvention propertyConvention,
             }
         }
 
-        private static Dictionary<Type, List<Type>> BuildDerivedTypesMapping(IWebApiAssembliesResolver assemblyResolver)
+        private static Dictionary<Type, Type[]> BuildDerivedTypesMapping(IWebApiAssembliesResolver assemblyResolver)
         {
-            IEnumerable<Type> allTypes = TypeHelper.GetLoadedTypes(assemblyResolver).Where(t => TypeHelper.IsVisible(t) && TypeHelper.IsClass(t) && t != typeof(object));
-            Dictionary<Type, List<Type>> allTypeMapping = allTypes.Distinct().ToDictionary(k => k, k => new List<Type>());
-
-            foreach (Type type in allTypes)
+            Predicate<Type> verifyType = t => TypeHelper.IsVisible(t) && TypeHelper.IsClass(t) && t != typeof(object);
+
+            // The dictionary is allocated to contain all visible reference types which is not system.Object, each with an List<Type> allocated.
+            // The list is used to keep derived types.
+            // In a common scenario, there are about 10% types with derived types.
+            // So, we don't need to keep a large number of empty list in the dictionary.
+            Dictionary<Type, List<Type>> temp = new Dictionary<Type, List<Type>>();
+            foreach (Type type in TypeHelper.GetLoadedTypes(assemblyResolver))
             {
-                List<Type> derivedTypes;
-                if (TypeHelper.GetBaseType(type) != null && allTypeMapping.TryGetValue(TypeHelper.GetBaseType(type), out derivedTypes))
+                if (type != null && verifyType(type))
                 {
-                    derivedTypes.Add(type);
+                    Type baseType = TypeHelper.GetBaseType(type);
+                    if (baseType != null)
+                    {
+                        List<Type> list;
+                        if (!temp.TryGetValue(baseType, out list))
+                        {
+                            if (verifyType(baseType))
+                            {
+                                list = new List<Type>(1);
+                                temp[baseType] = list;
+                            }
+                        }
+
+                        if (list != null)
+                        {
+                            list.Add(type);
+                        }
+                    }
                 }
             }
 
-            return allTypeMapping;
+            // We can throw away all lists and keep the dictionary as small as possible
+            Dictionary<Type, Type[]> map = new Dictionary<Type, Type[]>(temp.Count);
+            foreach (var kv in temp)
+            {
+                map[kv.Key] = kv.Value.ToArray();
+            }
```

Comment:
rather than copying to arrays to reduce unused space, could you just use temp.TrimExcess()?

Reply:
That's the suggestion from Feng. "As such data structure is normally generated once and kept in memory forever, I would event change it to use array. Now we’re throwing away all lists and keep the dictionary minimum. Notice here ToArray is a real method on List, not LINQ extension method"


---
In reply to: [598868250](https://github.com/OData/WebApi/pull/2436#discussion_r598868250) [](ancestors = 598868250)

Reply:
Looks like TrimExcess() does a reallocate/copy, so ToArray() is likely a better alternative.

---
In reply to: [598894104](https://github.com/OData/WebApi/pull/2436#discussion_r598894104) [](ancestors = 598894104,598868250)